### ISSUE TITLE
Debugger: Smart Step Into

### ISSUE THREAD

**Issue by hediet** [microsoft/vscode#123879]
# Problem

Given that we are debugging the following code:

```ts
console.log(➡️add(➡️f1(), ➡️f2())); // This is the currently active line that is about to be run.

function f1() {
    // ...
    return 1;
}

function f2() {
    // ...
    return 2;
}

function add(a, b) {
    return a + b + 1;
}
```

Next, we want to step into `add`. However, if we use the "Step Into" command, we step into `f1` and `f2` first.
This is especially annoying when there are a lot of uninteresting `getXYZ` functions/getters.

# Suggested Solution

It would be nice if VS Code had a way so that the user can specify in which method/function they want to step into.
[IntelliJ](https://www.jetbrains.com/help/idea/stepping-through-the-program.html#smart-step-into) and [PhpStorm](https://blog.jetbrains.com/phpstorm/2013/07/debugging-with-smart-step-into/) have a "smart step into" feature that does this.

If there is only one target to step into, nothing can be done.
However, if there are multiple targets that the debugger can step into, it would be nice if the user could select one.

There a many ways how the selection story can be implemented.

* **By using a quick pick dialog.** ([like PhpStorm in 2013](https://blog.jetbrains.com/phpstorm/2013/07/debugging-with-smart-step-into/)) If "Step Into" is called and the debug adapter reports multiple targets, show them all in a quick pick dialog (it would list "add", "f1" and "f2"). If the user selects one, the debugger will step into the selected target. This might be confusing if a method appears twice.

* **Source Code Range Highlight + Selection.** (similar to [IntelliJ](https://www.jetbrains.com/help/idea/stepping-through-the-program.html#smart-step-into)) If "Step Into" is called and the debug adapter reports multiple target ranges, highlight all those ranges like this (imagine the line being yellow):
![image](https://user-images.githubusercontent.com/2931520/118259971-125a0580-b4b2-11eb-906a-16de1a2c78b3.png)
If the user clicks on a range, the debugger will step into the selected target. With the arrow keys to select a target range and enter to accept it, a keyboard only story can be provided.

* **Modifier Key + Inline Decorators.**
While the user holds `Ctrl`, inline target decorators are shown before each step-into target in the line that is about to be executed like this (imagine the line being yellow):
![image](https://user-images.githubusercontent.com/2931520/118261965-8695a880-b4b4-11eb-9f26-afc576ea2c3b.png)
If the user clicks on one of those decorators while `Ctrl` is pressed, the debugger steps into the selected target. "Step Into" does not need to be invoked before for this scenario and does not need to be changed. This still allows for the "Go To Declaration/Definition" scenario, as separate click targets are shown.
This could be extended for "Go To Cursor" too ([similar to how Chrome does it](https://developer.chrome.com/blog/new-in-devtools-60/#continue), except that they don't render additional inline decorators as they don't have the "Go To Definition" feature).
This approach is very efficient for users who debug with the mouse. Keyboard users could enter this "Show All Step-Into Target Decorators" mode with a specific command and then use the arrow keys to select a target.

# JS Debug Adapter Implementation Details

Afaik, the V8 debugger does not directly provide the information/commands required for this.
Some heuristic involving the AST might be needed. I think it is ok if this feature is not enabled in situations where the list of available targets cannot be statically computed. If it is enabled though, it should always work.

The JS debug adapter could always "step in" and use "step out" if the method that V8 stepped into does not match the selected target.

**Comment by isidorn** [comment]
@hediet thanks a lot for creating this issue and for writing down the ideas.
For now assigning also to @connor4312 
fyi @weinand

**Comment by connor4312** [comment]
Hm, yea. This is tricky to do. An AST heuristic would not necessarily work, since you can step into getter functions which are indistinguishable from property accesses at the AST level. Though we could combine an AST heuristic and "ignore" getter functions, that has potential to work.

For named functions we can step in and step out until we pause at a function matching the name, which would work for many cases but would not handle any more intricate redirection or anonymous functions. I wonder how jetbrains does it.

Another idea would be using the language service's "Go to Implementation" with a Run to Cursor-like mechanism. This would work across languages, but fairly poor reliability.

I'm guessing Jetbrains does the AST approach.

**Comment by hediet** [comment]
I don't know if JetBrains does it for JavaScript, they do it for C#, Java and PHP though.
(Also, this feature has two sides: The VS Code UI and the JavaScript debugger - however, having the UI does not make much sense if there is no debug adapter that implements it)

Btw., you can distinguish getter functions from property accesses by using CDB. There is a function for side-effect free evaluation.

I think it would already be enough if it works for simple functions/methods that are not renamed (like, I would say, it is usually the case). In all other cases, this feature could just be not available, so that there will be no "buggy" behavior that users can complain about.

**Comment by connor4312** [comment]
Ok, so for JS we could actually do this quite well. The call frame when we step into arguments actually gives us the exact column the function is being called from, so we can remember where functions in that line are and step in and out until we get to the right one. We will just need to do some AST parsing, which might fail for certain TypeScript functions but should work generally well (we use [acorn's loose mode](https://github.com/acornjs/acorn) for general parsing in js-debug)

**Comment by bpasero** [comment]
I had filed https://github.com/microsoft/vscode/issues/119491 a while ago: I really like the feature where I can Ctrl+click into a method and the debugger stops in the first line of that method. This would also not add any breakpoints it would just stop once. This way you can navigate around the code with the debugger.

**Comment by weinand** [comment]
VS Code supports this feature already via DAP's "StepInTargets" Request.

But debug extensions must opt into this (Mock Debug does, I'm not sure about C# and C++)

**Comment by weinand** [comment]
@bpasero I like your proposed feature #119491 too, but that is not covered by the mentioned VS Code's "StepInTargets" feature.

**Comment by connor4312** [comment]
Oh, neat, I've never seen that! I will track this as a feature request for js-debug and re-tag Isi if I find any missing editor integration.

**Comment by hediet** [comment]
@weinand I guess that is a separate issue and also depends on the success of a possible implementation in js-debug, but what are your thoughts on adding a source code range to StepInTarget? Even if not used today, but that would enable a very nice UI one day.

**Comment by weinand** [comment]
@hediet yes, a `range` property on the `StepInTarget` type was already on the radar when we discussed the DAP request initially. But since we (VS Code) were too far away from adopting of the `StepInTarget` feature in VS Code, we decided not to add any properties with unclear UI representation.

Please create a new feature request in the DAP repo.

**Comment by hediet** [comment]
@isidorn and I did a quick brainstorming session together. 

We had a nice idea, however the following gif shows how difficult it is:

![recording](https://user-images.githubusercontent.com/2931520/124157295-e82bc980-da98-11eb-89fe-c4f209321bb7.gif)

Basically, for the `bar` marked in yellow, chrome dev tools don't allow to set a breakpoint. This means you really have to set the breakpoint in `bar` itself if you want to step into that particular invocation of `bar`. This means that some smartness is required to see where a function invocation will jump to.

Otherwise, you could implement "Step into Target" by adding a temporary breakpoint to the target expression and then stepping into it.

**Comment by connor4312** [comment]
In the approach in https://github.com/microsoft/vscode/issues/123879#issuecomment-841393181, I suggested doing a repeated step in and step out until we get to where we need to be. As you found, setting breakpoints in complex expressions can be hit and miss, but I think this should work given that line and column numbers of the callers are reported correctly.

**Comment by Trass3r** [comment]
> VS Code supports the proposed "Smart Step Into" feature already via DAP's [StepInTargets](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_StepInTargets) Request.
> 
> But debug extensions must opt into this (Mock Debug does, js-debug does not yet, I'm not sure about C# and C++)

The C++ ticket is https://github.com/microsoft/vscode-cpptools/issues/2488
But it'd require compiler support according to the gdb bug ticket.

**Comment by akaroml** [comment]
Here to upvote for the second solution mentioned by @hediet 

When debugging, to step into a target, we now need to right-click on a symbol and pick "Step into target" from the context menu. This works, but the context menu entry has a weakness of discoverability, and the context menu is also a distraction.

With the second solution, the "Step into target" melts into the regular "Step into," while visual hints are presented to the users directly. The benefits are
1. The capability to step into a target is more easily discovered, and no extra learning is needed
2. Less steps required
3. No visual distraction (no context menu involved)

While the implementation of debuggers (Java debugger AFAIK) already reports the target candidates to vscode, it should be feasible now.

**Comment by FeldrinH** [comment]
I also support the seconds solution mentioned in the original post.

In my experience there are three things that make the current approach clumsy which the second solution solves:
1. Having to manullay open a context menu to select the step in target gets quite tedious if you do it a lot. 
2. Having separate step into and step into targets actions makes it possible to accidentally step into the default target without intending to.
3. A lot of the times the labels provided by the debuggers I use aren't very clear. Having a visual indicator of the location of the step in target would be make it much clearer what I am selecting.
